adapter/internal/discovery/xds/rate_limiter_cache.go

```diff
@@ -0,0 +1,253 @@
+			if _, ok := apiOperations[path]; ok {
+				// Unreachable if the swagger definition is valid
+				loggers.LoggerXds.Warnf("Duplicate API resource %q in the swagger definition, skipping rate limit policy for the duplicate resource.", path)
+				continue
+			}
+
+			apiOperations[path] = make(map[string]struct{})
+			operationRlsConfigs := make([]*rls_config.RateLimitDescriptor, 0, len(resource.GetMethod()))
+			operationPolicyExists := false
+			for _, operation := range resource.GetMethod() {
+				method := operation.GetMethod()
+				if _, ok := apiOperations[path][method]; ok {
+					// Unreachable if the swagger definition is valid
+					loggers.LoggerXds.Warnf("Duplicate API resource HTTP method %q %q in the swagger definition, skipping rate limit policy for the duplicate resource.", path, method)
+					continue
+				}
+
+				if operation.RateLimitPolicy == "" {
+					loggers.LoggerXds.Debugf("No Operation level rate limit policy defined for %q %q", path, method)
+					continue
+				}
+				operationPolicyExists = true
+
+				rlPolicyConfig, err := getRateLimitPolicy(policies, operation.RateLimitPolicy)
+				if err != nil {
+					loggers.LoggerXds.Errorf("Error generating rate limit configuration for API: %q, API operation: %q %q: %v",
+						apiID, path, method, err)
+					return err
+				}
+
+				rlConf := &rls_config.RateLimitDescriptor{
+					Key:       "method",
+					Value:     method,
+					RateLimit: rlPolicyConfig,
+				}
+				operationRlsConfigs = append(operationRlsConfigs, rlConf)
+				apiOperations[path][method] = void
+			}
+
+			if operationPolicyExists {
+				rlsConfig := &rls_config.RateLimitDescriptor{
+					Key:         "path",
+					Value:       path,
+					Descriptors: operationRlsConfigs,
+				}
+				rlsConfigs = append(rlsConfigs, rlsConfig)
+			}
+		}
+	} else {
+		return fmt.Errorf("invalid rate limit policy level: %q", level)
+	}
+
+	org := mgwSwagger.OrganizationID
+	vHost := mgwSwagger.VHost
+
+	if _, ok := r.apiLevelRateLimitPolicies[org]; !ok {
+		r.apiLevelRateLimitPolicies[org] = make(map[string]map[string][]*rls_config.RateLimitDescriptor)
+	}
+
+	if _, ok := r.apiLevelRateLimitPolicies[org][vHost]; !ok {
+		r.apiLevelRateLimitPolicies[org][vHost] = make(map[string][]*rls_config.RateLimitDescriptor)
+	}
+
+	r.apiLevelRateLimitPolicies[org][vHost][apiID] = rlsConfigs
+	return nil
+}
+
+// DeleteAPILevelRateLimitPolicies deletes inline Rate Limit policies added with the API.
+func (r *rateLimitPolicyCache) DeleteAPILevelRateLimitPolicies(org, vHost, apiID string) {
+	delete(r.apiLevelRateLimitPolicies[org][vHost], apiID)
+}
+
+func getRateLimitPolicy(policies map[string]*mgw.APIRateLimitPolicy, policyName string) (*rls_config.RateLimitPolicy, error) {
+	policy, ok := policies[policyName]
+	if !ok {
+		return nil, fmt.Errorf("rate limit policy %q not defined", policyName)
+	}
+
+	unit, err := getRateLimitUnit(policy.SpanUnit)
+	if err != nil {
+		return nil, err
```

Comment:
Can we add a debug log in here ?

Reply:
Error is passed to the caller, and caller log the error.